src/libopensc/card-piv.c

```diff
@@ -2031,14 +2031,16 @@ static int piv_get_serial_nr_from_CHUI(sc_card_t* card, sc_serial_number_t* seri
 				/* test if guid and the fascn starts with ;9999 (in ISO 4bit + partiy code) */
 				if (!(gbits && fascn[0] == 0xD4 && fascn[1] == 0xE7
 						    && fascn[2] == 0x39 && (fascn[3] | 0x7F) == 0xFF)) {
-					serial->len = fascnlen < SC_MAX_SERIALNR ? fascnlen : SC_MAX_SERIALNR;
+					/* fascnlen==25 < SC_MAX_SERIALNR */
+					serial->len = fascnlen;
 					memcpy (serial->value, fascn, serial->len);
 					r = SC_SUCCESS;
 					gbits = 0; /* set to skip using guid below */
 				}
 			}
 			if (guid && gbits) {
-				serial->len = guidlen < SC_MAX_SERIALNR ? guidlen : SC_MAX_SERIALNR;
+				/* with gbits != 0 guidlen==16 < SC_MAX_SERIALNR */
+				serial->len = guidlen;
 				memcpy (serial->value, guid, serial->len);
 				r = SC_SUCCESS;
 			}
```

Comment:
The original code was still correct. 

The change appears to make the assumption the fascnlen  (25) is always less then SC_MAX_SERIALNR (32) which is true as long as  the difinition of SC_MAX_SERIALNR does not change. 
If it did, there is a chance of over writing in line 2036.  

Same goes for guidlen (16) and SC_MAX_SERIALNR  and overwrite in line 2044. 

If SC_MAX_SERIALNR was to change, would  Conerity  catch the possible overwrite?
Does a good optimizing compiler optimize out the code?